# **Walmart Sales Analysis**

### Project Overview

This project focuses on performing end-to-end data analysis on Walmart’s transactional sales dataset using Python, SQL, and data-cleaning best practices.

The goal is to:

- Explore and understand the structure of the dataset

- Identify data quality issues

- Clean and preprocess the data for SQL loading

- Prepare the dataset for answering business-critical questions

- Build a reproducible, portfolio-friendly analysis workflow
---


## **Step 1 - Initial Data Exploration**

#### 1.1 **Import Required Libraries**

In [ ]:
import pandas as pd

import pymysql  # MySQL database connector
from sqlalchemy import create_engine  # Creates SQLAlchemy database engine

#### 1.2 **Load dataset**

In [141]:
df = pd.read_csv('Walmart.csv')

#### 1.3 **Dataset Snapshot (head)**

In [142]:
df.head()

invoice_id   Branch         City                category unit_price  \
0           1  WALM003  San Antonio       Health and beauty     $74.69   
1           2  WALM048    Harlingen  Electronic accessories     $15.28   
2           3  WALM067  Haltom City      Home and lifestyle     $46.33   
3           4  WALM064      Bedford       Health and beauty     $58.22   
4           5  WALM013       Irving       Sports and travel     $86.31   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48

#### **Observation:**

The first few rows show expected fields:
invoice_id, Branch, City, category, unit_price, quantity, date, time, payment_method, rating, profit_margin.

Notice:

- unit_price includes $ symbol -> string, needs cleaning
---

#### 1.4 **Shape of the Dataset**

In [143]:
df.shape

(10051, 11)

#### 1.5 **Column Names**

In [144]:
df.columns

Index(['invoice_id', 'Branch', 'City', 'category', 'unit_price', 'quantity',
       'date', 'time', 'payment_method', 'rating', 'profit_margin'],
      dtype='object')

#### **Observation:**

- Column names look clean and well-structured.
- We will later convert them to lowercase for consistency before SQL export.

---

#### 1.6 **Dataset Information (dtypes + non-null counts)**

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_id      10051 non-null  int64  
 1   Branch          10051 non-null  object 
 2   City            10051 non-null  object 
 3   category        10051 non-null  object 
 4   unit_price      10020 non-null  object 
 5   quantity        10020 non-null  float64
 6   date            10051 non-null  object 
 7   time            10051 non-null  object 
 8   payment_method  10051 non-null  object 
 9   rating          10051 non-null  float64
 10  profit_margin   10051 non-null  float64
dtypes: float64(3), int64(1), object(7)
memory usage: 863.9+ KB


#### **Observations:**

- `unit_price` is object/string, contains $, needs conversion -> numeric.

- `date` and `time` columns are stored as strings, need conversion -> datetime.

- `quantity` has 31 missing values.
---


#### 1.7 **Summary Statistics**

In [146]:
df.describe()

invoice_id      quantity        rating  profit_margin
count  10051.000000  10020.000000  10051.000000   10051.000000
mean    5025.741220      2.353493      5.825659       0.393791
std     2901.174372      1.602658      1.763991       0.090669
min        1.000000      1.000000      3.000000       0.180000
25%     2513.500000      1.000000      4.000000       0.330000
50%     5026.000000      2.000000      6.000000       0.330000
75%     7538.500000      3.000000      7.000000       0.480000
max    10000.000000     10.000000     10.000000       0.570000

#### **Observations:**


- The `invoice_id` column shows 10,051 entries, while the maximum invoice ID value is 10,000. This suggests that some invoice IDs appear more than once in the dataset, indicating repeated invoice records. Further analysis is required to determine whether these repetitions represent multiple product line items under the same invoice or exact duplicate rows.

- The `quantity` column has 10,020 non-null values, indicating 31 missing entries. Quantity ranges from 1 to 10, with a median of 2 and an average of ~2.35. The standard deviation shows **low variation**, meaning most customers buy similar small quantities.

- The `rating` column contains 10,051 entries, with ratings from 3 to 10. The median is 6 and the average is ~5.82. The standard deviation suggests **moderate spread**, meaning ratings vary but still lean toward the mid–higher side.

- The `profit_margin` column has 10,051 non-null entries, ranging from 0.18 to 0.57. The median is 0.33 and the average is ~0.39. The standard deviation indicates **moderate variability**, meaning profit margins fluctuate but stay within a controlled range.


---

#### 1.8 **Unique Values in invoice_id**

In [147]:
df['invoice_id'].nunique()

10000

#### 1.9 **Duplicate Check**

In [148]:
df.duplicated().sum()

np.int64(51)

- This confirms `invoice_id` has 10,000 unique values, and 51 exact duplicate rows exist --> These must be removed
---

#### 1.10 **Missing Values Summary**

In [149]:
df.isnull().sum()

invoice_id         0
Branch             0
City               0
category           0
unit_price        31
quantity          31
date               0
time               0
payment_method     0
rating             0
profit_margin      0
dtype: int64

#### **Observations:**
- Only unit_price and quantity have missing values (31 each).

- All other fields are complete.

- Missingness is very small (31/10051 = 0.3%), safe to impute or drop.
---

#### 1.11 **Cardinality Check (Unique Values Across Columns)**

In [150]:
df.nunique()

invoice_id        10000
Branch              100
City                 98
category              6
unit_price         1008
quantity             10
date               1460
time               1001
payment_method        3
rating               62
profit_margin         6
dtype: int64

#### 1.12 **Category Distribution**

In [151]:
df['category'].value_counts()

category
Fashion accessories       4579
Home and lifestyle        4561
Electronic accessories     419
Food and beverages         174
Sports and travel          166
Health and beauty          152
Name: count, dtype: int64

#### 1.13 **Payment Method Distribution**

In [152]:
df['payment_method'].value_counts()

payment_method
Credit card    4260
Ewallet        3911
Cash           1880
Name: count, dtype: int64

#### 1.14 **City Distribution**

In [153]:
df['City'].value_counts()

City
Weslaco            399
Waxahachie         381
Port Arthur        240
Plano              238
Richardson         233
                  ... 
Irving              57
Lewisville          56
College Station     56
Amarillo            52
Lake Jackson        51
Name: count, Length: 98, dtype: int64

#### 1.15 **Branch Distribution**

In [154]:
df['Branch'].value_counts()

Branch
WALM058    240
WALM009    238
WALM030    233
WALM069    224
WALM074    212
          ... 
WALM013     57
WALM031     56
WALM034     56
WALM014     52
WALM092     51
Name: count, Length: 100, dtype: int64

---
## **Step 2 — Data Cleaning & Feature Engineering**

#### 2.1 **Remove Duplicate Rows**

In [155]:
# Remove exact duplicate rows

df = df.drop_duplicates()

df.shape

(10000, 11)

#### 2.2 **Clean unit_price Column (Remove $ and Convert to Float)**

In [156]:
# Clean unit_price column
df['unit_price'] = df['unit_price'].str.replace('$', '', regex=False)
df['unit_price'] = pd.to_numeric(df['unit_price'], errors='coerce')

- $ symbol removed successfully.

- All unit prices now numeric (float).

- Any invalid values automatically become NaN (none expected except originally missing 31 rows).
---


#### **2.3 Convert quantity to Integer**

In [157]:
df['quantity'] = df['quantity'].astype('Int64')   # allows integer + NaN

- Quantity is integer-based.

- Using `Int64` supports missing values correctly.
---

#### 2.4 **Handle Missing Values (unit_price & quantity)**

> Only these two columns have missing values.
Best practice: Remove rows with both missing because sales cannot be calculated without quantity or price.

In [158]:
df = df.dropna(subset=['unit_price', 'quantity'])
df.shape

(9969, 11)


- 31 rows removed (where both values were missing together).

- Clean dataset now has: 9,969 rows.

- No imputation required.
---

#### 2.5 **Convert `date` Column to Datetime**

In [159]:
df['date'] = pd.to_datetime(df['date'], format="%d/%m/%y")

- Date column now recognized as proper datetime.

- Enables extraction of year, month, day, weekday.
---

#### 2.6 **Convert `time` Column to Datetime (Time Only)**

In [160]:
df['time'] = pd.to_datetime(df['time'], format='%H:%M:%S').dt.time

- Time stored as Python time object.

- Will extract hour next for shift-based analysis.
---

#### 2.7 **Create total_amount Feature**

In [161]:
df['total_amount'] = df['unit_price'] * df['quantity']

- Essential metric: revenue per transaction.

- Required for nearly all business questions.
---

#### 2.10 **Normalize Column Names**

In [162]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
df.columns

Index(['invoice_id', 'branch', 'city', 'category', 'unit_price', 'quantity',
       'date', 'time', 'payment_method', 'rating', 'profit_margin',
       'total_amount'],
      dtype='object')

- Ensures SQL-friendly naming.

- Avoids issues with spaces and capitals in PostgreSQL/MySQL.

In [163]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9969 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   invoice_id      9969 non-null   int64         
 1   branch          9969 non-null   object        
 2   city            9969 non-null   object        
 3   category        9969 non-null   object        
 4   unit_price      9969 non-null   float64       
 5   quantity        9969 non-null   Int64         
 6   date            9969 non-null   datetime64[ns]
 7   time            9969 non-null   object        
 8   payment_method  9969 non-null   object        
 9   rating          9969 non-null   float64       
 10  profit_margin   9969 non-null   float64       
 11  total_amount    9969 non-null   Float64       
dtypes: Float64(1), Int64(1), datetime64[ns](1), float64(3), int64(1), object(5)
memory usage: 1.0+ MB


In [164]:
df.head()

invoice_id   branch         city                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity       date      time payment_method  rating  profit_margin  \
0         7 2019-01-05  13:08:00        Ewallet     9.1           0.48   
1         5 2019-03-08  10:29:00           Cash     9.6           0.48   
2         7 2019-03-03  13:23:00    Credit card     7.4           0.33   
3         8 2019-01-27  20:33:00        Ewallet     8.4           0.33   
4         7 2019-02-08  10:37:00        Ewallet     5.3           0.48   

   total_amount  
0        522.83  
1          76.4  
2        324.31  
3        465.76  
4        604.17

In [165]:
df.describe()

invoice_id   unit_price  quantity                           date  \
count   9969.000000  9969.000000    9969.0                           9969   
mean    5010.116561    50.622142  2.355602  2021-12-01 16:46:56.731868672   
min        1.000000    10.080000       1.0            2019-01-01 00:00:00   
25%     2524.000000    32.000000       1.0            2020-11-25 00:00:00   
50%     5016.000000    51.000000       2.0            2021-12-05 00:00:00   
75%     7508.000000    69.000000       3.0            2022-12-18 00:00:00   
max    10000.000000    99.960000      10.0            2023-12-31 00:00:00   
std     2886.217349    21.203766  1.605455                            NaN   

            rating  profit_margin  total_amount  
count  9969.000000    9969.000000        9969.0  
mean      5.828839       0.393744    121.348819  
min       3.000000       0.180000         10.17  
25%       4.000000       0.330000          54.0  
50%       6.000000       0.330000          88.0  
75%       7.000000       0.480000         156.0  
max      10.000000       0.570000         993.0  
std       1.763723       0.090659     112.67804

In [166]:
df.to_csv('Walmart_clean_data.csv', index=False)

In [167]:
#mysql connection
# "mysql+pymysql://user:password@localhost:3306/db_name"

engine_mysql = create_engine('mysql+pymysql://root@localhost:3306/walmart_db')

try:
    engine_mysql
    print('Connection Successful')
except:
    print('Connection Unsuccessful')

Connection Successful


In [169]:
df.to_sql(name='walmart', con=engine_mysql, if_exists='append', index=False)

9969